File: src/main/java/DahRealPanda/plugins/colonyweb/auth/WebUser.java
package DahRealPanda.plugins.colonyweb.auth;

import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;

/**
 * A Minecraft player's dashboard account: which colonies they may view, and their live
 * browser sessions.
 *
 * <p>Two sources of access are tracked separately so a re-sync never silently drops an
 * operator's explicit grant:</p>
 * <ul>
 *   <li>{@link #colonies} — mirrored from MineColonies permissions by {@code /colonyweb sync}
 *       and replaced wholesale on every sync.</li>
 *   <li>{@link #granted} — added by hand with {@code /colonyweb access grant} and only ever
 *       removed by {@code /colonyweb access revoke}.</li>
 * </ul>
 */
public class WebUser {
    public String uuid;
    public String name;

    /** Colonies the player belongs to in-game (owner/officer/member). Replaced on each sync. */
    public Set<Integer> colonies = new LinkedHashSet<>();

    /** Colonies an operator granted explicitly. Survives re-syncs. */
    public Set<Integer> granted = new LinkedHashSet<>();

    /** Server operators see every colony. */
    public boolean admin;

    /** Epoch millis of the last {@code /colonyweb sync}. */
    public long syncedAt;

    /** Active browser sessions. Only token hashes are stored, never the tokens themselves. */
    public List<StoredSession> sessions = new ArrayList<>();

    public WebUser() {
    }

    public WebUser(String uuid, String name) {
        this.uuid = uuid;
        this.name = name;
    }

    /** Every colony id this user may view (ignored when {@link #admin} is set). */
    public Set<Integer> accessibleColonies() {
        Set<Integer> all = new LinkedHashSet<>(colonies);
        all.addAll(granted);
        return all;
    }

    /** Remove sessions that have expired, and report whether anything changed. */
    public boolean purgeExpiredSessions(long now) {
        return sessions.removeIf(session -> session.expiresAt <= now);
    }
}

File: common/src/main/kotlin/DahRealPanda/plugins/colonyweb/auth/WebUser.kt
package DahRealPanda.plugins.colonyweb.auth

import java.util.LinkedHashSet

/**
 * A Minecraft player's dashboard account: which colonies they may view, and their live
 * browser sessions.
 *
 * Two sources of access are tracked separately so a re-sync never silently drops an
 * operator's explicit grant:
 *
 *  * [colonies] — mirrored from MineColonies permissions by `/colonyweb sync`
 *    and replaced wholesale on every sync.
 *  * [granted] — added by hand with `/colonyweb access grant` and only ever
 *    removed by `/colonyweb access revoke`.
 */
data class WebUser(
    var uuid: String = "",
    var name: String = "",
    var colonies: MutableSet<Int> = LinkedHashSet(),
    var granted: MutableSet<Int> = LinkedHashSet(),
    var admin: Boolean = false,
    var syncedAt: Long = 0L,
    var sessions: MutableList<StoredSession> = ArrayList()
) {
    fun accessibleColonies(): Set<Int> {
        val all = LinkedHashSet(colonies)
        all.addAll(granted)
        return all
    }

    /** Remove sessions that have expired, and report whether anything changed. */
    fun purgeExpiredSessions(now: Long): Boolean {
        return sessions.removeIf { session -> session.expiresAt <= now }
    }
}
